# IEEE-CIS Fraud — IID, группы и время

[IEEE-CIS Fraud Detection](https://www.kaggle.com/competitions/ieee-fraud-detection): бинарная классификация с `roc_auc`. Код объединяет таблицы транзакций и идентичности, затем сравнивает IID, group и time-series CV.

`uid = card1 + addr1 + D1` — составной групповой ключ. Раздельность таких ключей не доказывает раздельность реальных карт. Временной сценарий использует `TransactionDT`; group- и time-разбиения здесь проверяются отдельными fit, а не одной совместной схемой.

**Срез результатов: 9 сентября 2026.** Итоговые таблицы используют связанные отчёты контрольных прогонов: seed 42, четыре native/BLAS-потока, кеш выключен, HPO до пяти trials на семейство при его включении, отправки Kaggle выключены. Встроенные outputs относятся к собственному запуску ячеек; контрольные режимы baseline/fast запускались отдельным runner. Код ячеек сохраняет исходные бюджеты и не включает `SearchConfig`. Просмотренный holdout служит диагностикой. [Границы сравнения и независимой приёмки](README.md#контрольные-прогоны-и-интерпретация).

In [1]:
import json
import logging
import os
import shutil
import subprocess
import time
from pathlib import Path

import pandas as pd

from honestml import (
    AutoML,
    CVConfig,
    FeatureSelectionConfig,
    HPOConfig,
    render_report,
    save_run_report,
)

SEED = 42
SAMPLE_ROWS = 200_000  # contiguous early-time window; set to None for the full 590k rows
DATA = Path("data/ieee-fraud")
RESULTS = Path("results/ieee-fraud")
RESULTS.mkdir(parents=True, exist_ok=True)
KAGGLE = (
    os.environ.get("KAGGLE_BIN")
    or shutil.which("kaggle")
    or str(Path.home() / ".local" / "bin" / "kaggle")
)

logging.basicConfig(format="%(levelname)s %(name)s: %(message)s")
logging.getLogger("honestml").setLevel(logging.INFO)

In [2]:
if not (DATA / "train_transaction.csv").exists():
    DATA.mkdir(parents=True, exist_ok=True)
    subprocess.run(
        [
            KAGGLE,
            "datasets",
            "download",
            "-d",
            "lnasiri007/ieeecis-fraud-detection",
            "-p",
            str(DATA),
            "--unzip",
        ],
        check=True,
    )

In [3]:
tx = pd.read_csv(DATA / "train_transaction.csv")
idf = pd.read_csv(DATA / "train_identity.csv")
df = (
    tx.merge(idf, on="TransactionID", how="left")
    .sort_values("TransactionDT")
    .reset_index(drop=True)
)
if SAMPLE_ROWS is not None:
    df = df.iloc[:SAMPLE_ROWS].copy()

# the engineered client entity (the well-known 'magic uid'): same card recurs across transactions
uid = (
    df["card1"].astype("string").fillna("NA")
    + "_"
    + df["addr1"].astype("string").fillna("NA")
    + "_"
    + df["D1"].astype("string").fillna("NA")
).to_numpy()
tdt = df["TransactionDT"].to_numpy()  # the time axis (seconds from a fixed reference)
y = df.pop("isFraud")
X = df.drop(columns=["TransactionID", "TransactionDT"])  # ids/time axis are metadata, not features

n_groups = len(set(uid.tolist()))
rep = len(uid) / n_groups
print(f"rows: {len(X)}, features: {X.shape[1]}, fraud: {y.mean():.3%}")
print(f"uid groups: {n_groups} ({rep:.1f} transactions per card on average)")
X.head()

rows: 200000, features: 431, fraud: 3.012%
uid groups: 91869 (2.2 transactions per card on average)


,TransactionAmt,ProductCD,card1,card2,card3,card4,card5,card6,addr1,addr2,...,id_31,id_32,id_33,id_34,id_35,id_36,id_37,id_38,DeviceType,DeviceInfo
0,68.5,W,13926,NaN,150.0,discover,142.0,credit,315.0,87.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,29.0,W,2755,404.0,150.0,mastercard,102.0,credit,325.0,87.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,59.0,W,4663,490.0,150.0,visa,166.0,debit,330.0,87.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,50.0,W,18132,567.0,150.0,mastercard,117.0,debit,476.0,87.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,50.0,H,4497,514.0,150.0,mastercard,102.0,credit,420.0,87.0,...,samsung browser 6.2,32.0,2220x1080,match_status:2,T,F,T,T,mobile,SAMSUNG SM-G892A Build/NRD90M


## Run A — IID-разбиение

Стратифицированные folds и случайный holdout не ограничивают пересечение составных групп или порядок дат. Этот результат характеризует IID-сценарий. По разнице с другими CV нельзя отдельно оценить эффект запоминания карт или гарантировать качество на будущих транзакциях.

In [4]:
model_iid = AutoML(
    task="binary",
    metric="roc_auc",
    cv=CVConfig(outer_holdout=0.2, calibrate="auto"),
    random_state=SEED,
)
t0 = time.perf_counter()
model_iid.fit(X, y)
print(f"naive i.i.d. fit took {(time.perf_counter() - t0) / 60:.1f} min")
rep_iid = model_iid.run_report_
sel_iid = next(e["score"] for e in rep_iid["leaderboard"] if e["model_id"] == rep_iid["winner"])
print(
    f"winner {rep_iid['winner']}  selection roc_auc {sel_iid:.4f}  holdout {rep_iid['holdout_score']:.4f}  optimism {sel_iid - rep_iid['holdout_score']:+.4f}"
)
pd.DataFrame(rep_iid["leaderboard"])

INFO honestml.adapters.reader: auto-typing column=V107 dtype=Float64 role=ignore reason=constant


INFO honestml.adapters.reader: auto-typing column=id_33 dtype=String role=categorical reason=high_cardinality_string


INFO honestml.adapters.reader: auto-typing column=DeviceInfo dtype=String role=categorical reason=high_cardinality_string


WARNING honestml: native categorical gate demoted 4 high-cardinality column(s) to ordinal codes: ['id_30', 'id_31', 'id_33', 'DeviceInfo'] (native_cat_max_unique=64)


INFO honestml: stage key=run stage=selection elapsed=1197.0s


WARNING honestml.adapters.boosting: boosting 'lightgbm' trained without early stopping; leaderboard comparison may favor overfit settings


INFO honestml: stage key=run stage=refit elapsed=6.7s


INFO honestml: stage key=run stage=refit elapsed=7.7s


INFO honestml: stage key=run stage=finalize elapsed=7.7s


naive i.i.d. fit took 20.2 min
winner lightgbm  selection roc_auc 0.9515  holdout 0.9501  optimism +0.0015


,model_id,score,rank
0,lightgbm,0.951535,1
1,catboost,0.932963,2
2,linear,0.857911,3
3,baseline,0.499914,4


## Run B — разбиение по составному uid

`CVConfig(scheme="group")` и `groups=uid` разделяют строки по переданному ключу, включая outer holdout. Это проверка переноса между такими группами; семантику ключа и соответствие реальным сущностям нужно проверять отдельно. Другие даты и группы в test могут менять сложность задачи.

In [5]:
model_grp = AutoML(
    task="binary",
    metric="roc_auc",
    cv=CVConfig(scheme="group", outer_holdout=0.2, calibrate="auto"),
    random_state=SEED,
)
t0 = time.perf_counter()
model_grp.fit(X, y, groups=uid)
print(f"group-aware fit took {(time.perf_counter() - t0) / 60:.1f} min")
rep_grp = model_grp.run_report_
sel_grp = next(e["score"] for e in rep_grp["leaderboard"] if e["model_id"] == rep_grp["winner"])
print(
    f"winner {rep_grp['winner']}  selection roc_auc {sel_grp:.4f}  holdout {rep_grp['holdout_score']:.4f}  optimism {sel_grp - rep_grp['holdout_score']:+.4f}"
)
print(
    f"naive i.i.d. holdout was {rep_iid['holdout_score']:.4f}; group-honest holdout is {rep_grp['holdout_score']:.4f}"
)
pd.DataFrame(rep_grp["leaderboard"])

INFO honestml.adapters.reader: auto-typing column=V107 dtype=Float64 role=ignore reason=constant


INFO honestml.adapters.reader: auto-typing column=id_33 dtype=String role=categorical reason=high_cardinality_string


INFO honestml.adapters.reader: auto-typing column=DeviceInfo dtype=String role=categorical reason=high_cardinality_string


WARNING honestml: native categorical gate demoted 4 high-cardinality column(s) to ordinal codes: ['id_30', 'id_31', 'id_33', 'DeviceInfo'] (native_cat_max_unique=64)


INFO honestml: stage key=run stage=selection elapsed=1316.0s


INFO honestml: stage key=run stage=refit elapsed=6.8s


INFO honestml: stage key=run stage=refit elapsed=7.7s


INFO honestml: stage key=run stage=finalize elapsed=7.7s


group-aware fit took 22.2 min
winner lightgbm  selection roc_auc 0.9324  holdout 0.9494  optimism -0.0171
naive i.i.d. holdout was 0.9501; group-honest holdout is 0.9494


,model_id,score,rank
0,lightgbm,0.932350,1
1,catboost,0.919876,2
2,linear,0.847579,3
3,baseline,0.486898,4


## Run C — временной CV с FS и HPO

Код задаёт expanding folds по `TransactionDT`, purge/embargo по одному дню в секундах, `importance` FS и HPO. В ячейке предел составляет 15 trials/600 с; в контрольной серии — до пяти trials на семейство. Явный размер ES не задан.

Временной сценарий проверяет позднее окно. Он не гарантирует качество другого будущего периода или совместную независимость по времени и реальным картам. Поведение калибровки и отбора отражается в отчёте.

In [6]:
n_test = max(2000, len(y) // 12)  # rows per time fold — large enough for both classes at 3.5% fraud
one_day = (
    24 * 3600
)  # wall-clock purge/embargo: a fixed real duration, since TransactionDT is in seconds
model_ts = AutoML(
    task="binary",
    metric="roc_auc",
    preset="best",
    hpo=HPOConfig(n_trials=15, timeout_s=600),
    feature_selection=FeatureSelectionConfig(strategy="importance", cutoff="auto"),
    cv=CVConfig(
        scheme="timeseries",
        n_splits=5,
        n_test=n_test,
        purge_delta=one_day,
        embargo_delta=one_day,
        outer_holdout=0.2,
    ),
    random_state=SEED,
)
t0 = time.perf_counter()
model_ts.fit(X, y, time=tdt)
print(f"time-aware level-2 fit took {(time.perf_counter() - t0) / 60:.1f} min")

INFO honestml.composition.presets: preset best applied: ['ensemble']


INFO honestml.adapters.reader: auto-typing column=V107 dtype=Float64 role=ignore reason=constant


INFO honestml.adapters.reader: auto-typing column=id_33 dtype=String role=categorical reason=high_cardinality_string


INFO honestml.adapters.reader: auto-typing column=DeviceInfo dtype=String role=categorical reason=high_cardinality_string


WARNING honestml.composition.facade: cv.n_es=1 row(s) is a degenerate early-stopping tail for 157443 rows: stopping decisions are noise and boosting fits may run to the 1000-tree ES ceiling; set n_es to ~1-5% of the rows (>= 788 here)


WARNING honestml.application.feature_selection: feature selection kept 142 of 430 features


C:\Users\Admin\Documents\AutoML\.venv\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


INFO honestml: stage key=run stage=hpo elapsed=918.0s


WARNING honestml: native categorical gate demoted 4 high-cardinality column(s) to ordinal codes: ['id_30', 'id_31', 'id_33', 'DeviceInfo'] (native_cat_max_unique=64)


INFO honestml: stage key=run stage=selection elapsed=1421.3s


INFO honestml: stage key=run stage=ensemble elapsed=56.2s


INFO honestml: stage key=run stage=refit elapsed=5.6s


INFO honestml: stage key=run stage=refit elapsed=9.0s


INFO honestml: stage key=run stage=finalize elapsed=9.0s


time-aware level-2 fit took 24.9 min


In [7]:
rep_ts = model_ts.run_report_
sel_ts = next(e["score"] for e in rep_ts["leaderboard"] if e["model_id"] == rep_ts["winner"])
print(f"winner          : {rep_ts['winner']}")
print(f"selection score : {sel_ts:.4f}   (time-series OOF roc_auc)")
print(f"holdout score   : {rep_ts['holdout_score']:.4f}   (late-window holdout)")
print(f"optimism        : {sel_ts - rep_ts['holdout_score']:+.4f}")
print()
print("fs      :", json.dumps(rep_ts["feature_selection"], default=str)[:400])
print("hpo     :", json.dumps(rep_ts["hpo"], default=str)[:400])
print("ensemble:", json.dumps(rep_ts["ensemble"], default=str)[:300])
save_run_report(rep_ts, RESULTS)
md_path = render_report(rep_ts, RESULTS, fmt="md")
pd.DataFrame(rep_ts["leaderboard"])

winner          : lightgbm
selection score : 0.8643   (time-series OOF roc_auc)
holdout score   : 0.9025   (late-window holdout)
optimism        : -0.0382

fs      : {"strategy": "importance", "n_selected": 142, "selected": ["TransactionAmt", "card1", "card2", "card3", "card5", "addr1", "dist1", "dist2", "C1", "C2", "C4", "C6", "C7", "C8", "C9", "C10", "C11", "C12", "C13", "C14", "D1", "D2", "D3", "D4", "D5", "D8", "D9", "D10", "D11", "D15", "V12", "V13", "V19", "V20", "V24", "V35", "V36", "V37", "V38", "V44", "V45", "V47", "V52", "V53", "V54", "V56", "V61", "
hpo     : {"backend": "optuna", "inner_cv": 3, "deterministic": false, "selection_oof_is_post_tuning": true, "tuned_on_full_feature_space": false, "tuned_on": "fs_subset", "cost_estimate_fits": 90, "tuned": {"catboost": {"chosen_params": {"depth": 7, "learning_rate": 0.14447746112718687, "iterations": 100, "l2_leaf_reg": 3.2676417657817622, "subsample": 0.836965827544817, "one_hot_max_size": 4}, "inner_best
ensemble: {"applied": 

,model_id,score,rank
0,catboost,0.894457,1
1,lightgbm,0.864275,2
2,linear,0.817163,3
3,baseline,0.471704,4


## Результаты контрольной серии

| Сценарий | Режим | Итоговая модель | Fit, с | Просмотренный holdout |
|---|---|---|---:|---:|
| IID | baseline | lightgbm | 4662.66 | 0.955821 |
| IID | fast | catboost | 2339.41 | 0.936651 |
| Group | baseline | lightgbm | 3052.20 | 0.950596 |
| Group | fast | catboost | 2082.97 | 0.937859 |
| Time-series + level 2 | baseline | ансамбль (опорная lightgbm) | 1029.55 | 0.884460 |
| Time-series + level 2 | fast | lightgbm | 152.65 | 0.892786 |

Источник: [контрольная серия baseline/fast](../docs/training-results.md#comparison). Fast измерен с confirmation_rows=16384; это снимок серии, а не повтор всех сценариев на текущем коде библиотеки. Время относится к одному AutoML.fit.

В IID и group-сценариях frozen fast имеет меньшую viewed holdout ROC-AUC, чем baseline. Во временном расширенном сценарии направление обратное, но baseline поставляет ансамбль, fast — одиночный LightGBM. Различия CV и процедур не изолируют причину изменений качества.

[Baseline fit03](../docs/training-results.md#targeted) сохранил 142 признака и применил ансамбль четырёх моделей. Holdout ROC-AUC 0,8844597404 относится к ансамблю; DEV score 0,8507103459 — к выбранному одиночному LightGBM. Отдельный DEV score ансамбля в этом evidence не записан, поэтому эти две величины не образуют его optimism.

Для временного fast подтверждение отмечено `insufficient_confirmation`: ES содержит одну строку. [Отдельная IID DEV-серия](../docs/training-results.md#targeted) на 160000 DEV-строках и 430 признаках при confirmation_rows=65536 выбирает LightGBM на seeds 42/43/44. Она не оценивает holdout и не является новым полным AutoML-прогоном. Production-качество и независимая приёмка остаются открытыми.